### PR TITLE
bpo-42972: Track sqlite3 statement objects

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue42972

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @vstinner

File: Modules/_sqlite/cursor.c [hunk 1]
```diff
@@ -512,6 +512,7 @@ _pysqlite_query_execute(pysqlite_Cursor* self, int multiple, PyObject* operation
             goto error;
         }
         rc = pysqlite_statement_create(self->statement, self->connection, operation);
```
Comment (pablogsal) [inline Modules/_sqlite/cursor.c]: Tracking should be done on the constructor, not on the code calling it, as is not it's responsability. This way there is no need for synchronization and eliminates the possibility of getting a badly crated object.

**Review by pablogsal — CHANGES_REQUESTED** [review]

File: Modules/_sqlite/cursor.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: Yes, that's better. I'll move the allocation there as well, for consistency.

Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c]: I've moved allocation and GC tracking over to `pysqlite_statement_create`. PTAL.

File: Modules/_sqlite/statement.c [hunk 2]
```diff
@@ -48,29 +48,37 @@ typedef enum {
...
+    pysqlite_Statement *self = PyObject_GC_New(pysqlite_Statement,
+                                               pysqlite_StatementType);
+    if (self == NULL) {
+        return NULL;
+    }
+
+    self->st = NULL;
+    self->in_use = 0;
     self->in_weakreflist = NULL;
     self->sql = Py_NewRef(sql);
```
Comment (vstinner) [inline Modules/_sqlite/statement.c]: Could you initialize is_dml and db here?

File: Modules/_sqlite/statement.c [hunk 3]
```diff
@@ -48,29 +48,37 @@ typedef enum {
...
-    self->in_use = 0;
-
     assert(PyUnicode_Check(sql));
 
     sql_cstr = PyUnicode_AsUTF8AndSize(sql, &sql_cstr_len);
     if (sql_cstr == NULL) {
-        rc = PYSQLITE_SQL_WRONG_TYPE;
-        return rc;
+        PyErr_SetString(pysqlite_Warning,
+                        "SQL is of wrong type. Must be string.");
```
Comment (vstinner) [inline Modules/_sqlite/statement.c]: You may add Py_TYPE(sql)->tp_name in the error message.

```suggestion
                        "SQL is of wrong type ('%s'). Must be string.");
```

**Review by vstinner — APPROVED** [review]
LGTM, but I would like to see the minor issues being addressed first before merging your change.

File: Modules/_sqlite/statement.c (hunk 2 above)
Comment (erlend-aasland) [reply Modules/_sqlite/statement.c]: Sure! I tried to minimise the changes, but nothing would make me more glad than to refactor this a little bit more :)

Comment (erlend-aasland) [reply Modules/_sqlite/statement.c]: May I add a commit for intermingled variable declaration?

File: Modules/_sqlite/statement.c (hunk 3 above)
Comment (erlend-aasland) [reply Modules/_sqlite/statement.c]: BTW; this should probably have been a `TypeError`, not a `sqlite3.Warning`. Changing this is probably far out of scope for this PR.

File: Modules/_sqlite/statement.c [hunk 4]
```diff
@@ -48,35 +48,45 @@ typedef enum {
...
+    if (self == NULL) {
+        return NULL;
+    }
+
+    self->db = connection->db;
+    self->st = NULL;
     self->sql = Py_NewRef(sql);
+    self->in_use = 0;
+    self->is_dml = 0;
+    self->in_weakreflist = NULL;
```
Comment (vstinner) [inline Modules/_sqlite/statement.c:86]: Thanks, it looks better like that. It makes me me confident that calling Py_DECREF(self) is safe in this function.

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
Merged, thanks.

File: Modules/_sqlite/statement.c (hunk 4 above)
Comment (erlend-aasland) [reply Modules/_sqlite/statement.c:86]: Yes, this is way better. Thanks for pushing the PR in this direction.

**Comment by erlend-aasland** [comment]
> Merged, thanks.

Thanks for reviewing, Victor & Pablo!

**Comment by vstinner** [comment]
@erlend-aasland: Automated backport to 3.10 failed, can you try to backport the change manually?

@pablogsal: I'm not sure if this change is really a bugfix or an enhancement. I'm ok to backport it even after beta2 release.

**Comment by erlend-aasland** [comment]
> @erlend-aasland: Automated backport to 3.10 failed, can you try to backport the change manually?

Will do!

**Pull request merged**